ui/src/lib/fakeAPI.js

```
@@ -0,0 +1,71 @@
+  let users = [
+    {
+      id: 1,
+      username: 'simon',
+      password: 'simon',
+      firstName: 'Simon',
+      lastName: 'Ansell',
+    },
+  ]
+
+  let realFetch = window.fetch
+
+  window.fetch = function (url, opts) {
+    return new Promise((resolve, reject) => {
+      setTimeout(() => {
+        if (url.endsWith('/login') && opts.method === 'POST') {
+          let params = JSON.parse(opts.body)
+
+          let filteredUsers = users.filter(user => {
+            return (
+              user.username === params.username &&
+              user.password === params.password
+            )
+          })
+
+          if (filteredUsers.length) {
+            let user = filteredUsers[0]
+            let responseJson = {
+              id: user.id,
+              username: user.username,
+              firstName: user.firstName,
+              lastName: user.lastName,
+            }
+            resolve({
+              ok: true,
+              text: () => Promise.resolve(JSON.stringify(responseJson)),
+            })
+          } else {
+            reject('Username or password is incorrect.')
+          }
+
+          return
+        }
+
+        // example protected endpoint
+        if (url.endsWith('/users') && opts.method === 'GET') {
+          // check for fake auth token in header and return users if valid, this is implemented server side in a real application
+          if (
+            opts.headers &&
+            opts.headers.Authorization === `Basic ${window.btoa('simon:simon')}`
+          ) {
+            resolve({
+              ok: true,
+              text: () => Promise.resolve(JSON.stringify(users)),
+            })
+          } else {
+            // return 401 not authorised if token is null or invalid
+            resolve({ status: 401, text: () => Promise.resolve() })
+          }
+
+          return
+        }
+
+        // Requests not handled above
+        realFetch(url, opts).then(response => resolve(response))
```

this appears to silently discard failures?

---

It does, but see my comments above. `fakeAPI.js` is purely for development purposes when working without any actual endpoints and won't make it into production.